[WIP] Bitmask-based change tracking

Rich-Harris opened the pull request:
Another crack at #1943. First task is to confirm that it's possible and see how it affects the output; after that, need to figure out how to handle components with more than 31 changeable top-level values

stalkerg commented:
>  how to handle components with more than 31 changeable top-level values

Add another one bitmask? At least, PostgresDB does the same things for nulls in tuples.

tomcon commented:
Maybe using bigint:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt 
https://v8.dev/features/bigint
or characters?

Rich-Harris commented:
BigInt isn't really an option, it's not supported widely enough. One day. The most likely option at this point is an array of bitmasks

Rich-Harris commented:
Getting there. All the `runtime` tests are passing, but there are a few `js` tests where weird-looking code is being generated. Also, I carelessly broke sourcemaps, which will take a little bit of unpicking.

tivac commented on test/js/samples/bind-online/expected.js:
```diff
@@ -11,14 +11,11 @@ import {
-				listen(window, "online", ctx.onlinestatuschanged),
-				listen(window, "offline", ctx.onlinestatuschanged)
-			];
+			dispose = [listen(window, "online", ctx[1]), listen(window, "offline", ctx[1])];
```
It doesn't *really* matter, but the linebreaks were nice here.

Rich-Harris replied:
code-red breaks the line when it gets 'too long' which is a bit of a hand-wavey calculation (because it doesn't know, when calculating it, how much indentation will eventually be applied) — it would be easy enough to reduce the magic number involved in that calculation though

tivac replied:
Not convinced it's worth it, just calling out that the new version is a little awkward,

Also not gonna claim that I spend a lot of time looking at our generated JS and nitpicking the style of it.

Rich-Harris commented:
dumped some half-finished changes into a new branch, https://github.com/sveltejs/svelte/tree/bitmasks-2, because i *think* i've mostly done the right thing but i'm not certain yet and there are some tests failing that i don't have time to address right now. programming is hard.

Rich-Harris commented:
So, here's an interesting thing, that I suppose I should have realised beforehand: if a component *does* have more than 31 writable values, this approach still works. It just means that some things may get 'updated' that don't need to be (i.e. if the 35th variable is updated, any references to the 4th variable will also get updated, unnecessarily).

As long as the behaviour is correct, updating too frequently is relatively harmless. So I'm inclined to think maybe we *don't* need to worry about maintaining multiple bitmasks for those rare cases? Any thoughts?

Conduitry commented:
Huh. I assume this would also affect people who are using reactive blocks for situations where they want side effects each time something changes? That's probably not great.

Rich-Harris commented:
That's true. I suppose I'd put these false positives in the same category as the false positives you get on `obj.bar` when `obj.foo` is updated, but perhaps they're not?

Conduitry commented:
I suppose the question boils down to whether we are okay with telling people that their reactive blocks ought to be idempotent, and whether starting to tell people that is okay to happen in something that's not a major breaking release. I don't feel super comfortable answering 'yes' to those questions.

Rich-Harris commented:
What if we printed a warning in situations where you a) have more than 31 dynamic values and b) reactive blocks?

> Due to technical limitations, the following code...
> 
>      $: a = b * 2
>
> ...will re-run when `b` changes or when `z` changes. Consider breaking your component into two smaller components

Or is that a massive cop-out?

Conduitry commented:
I get that you really really don't want to fuck with arrays of bitmasks, and I'm not telling you that that's something you need to do, but it does feel like a little bit of a cop-out.

Rich-Harris commented:
It shouldn't be *that* big a deal as far as the compiled code is concerned — everything is fairly centralised. It just means we need to figure out what happens here:

* https://github.com/sveltejs/svelte/blob/3d8bd5ddad2ef42a6b8bb8441bb7a223f9e2264f/src/runtime/internal/Component.ts#L91
* https://github.com/sveltejs/svelte/blob/3d8bd5ddad2ef42a6b8bb8441bb7a223f9e2264f/src/runtime/internal/Component.ts#L124
* https://github.com/sveltejs/svelte/blob/3d8bd5ddad2ef42a6b8bb8441bb7a223f9e2264f/src/runtime/internal/scheduler.ts#L77

mrkishi commented:
As I see it, the more straightforward conversion would be to make dirty an array. Maybe a `(boolean, number[])` tuple would perform better, but we'd still be resetting the array somehow... So what I'd try first is:

- Initialize dirty to empty array, `dirty: []`.
- Adjust `make_dirty` accordingly:
```js
// i and j could be calculated at runtime from index,
// with i = (index / 32) | 0 and j = index % 32,
// but there's no reason to
function make_dirty(component, i, j) {
	if (component.$$.dirty.length === 0) {
		dirty_components.push(component);
		schedule_update();
	}
	component.$$.dirty[i] |= (1 << j);
}
```
- Reset the array: `$$.dirty = [];` or `$$.dirty.length = 0;`.

`ctx[index]` should still work, just `dirty`/`changed` that would be divided in buckets.

Performance _could_ be better with a `boolean` instead of testing for `length`, and optimizers _could_ perform better for usual less-than-too-many-vars components if `dirty` was initialized to `[0,0,0]`, but that'd need some good practical tuning.

Rich-Harris commented:
It's definitely an option. It would mean that every reference to `#changed` has to become `#changed[0]` which seems a shame. May be simpler than the alternative though.

mrkishi commented:
`#changed` is only used in generated code, right? Is the only interaction point with runtime code in `dirty`?

Would it be a pain to only use the extra `dirty` buckets when over 31 variables? As in, `$$.fragment && $$.fragment.p($$.ctx, $$.dirty[0], $$.dirty);`

Rich-Harris commented:
Yep (and it would probably make sense to use the same word in both cases).

I suppose one thing you could do is in the main fragment, have an update function like this in the <31 case...

```js
return {
  // ...
  p(ctx, ...changed) {
    child_block.p(ctx, changed);
  }
};
```

...rather than `changed` being passed around as an array. In the >=31 case it'd be as it is now:

```js
return {
  // ...
  p(ctx, changed) {
    child_block.p(ctx, changed);
  }
};
```

Conduitry commented:
Sometime in the past when discussing bitmasks, we'd mention having comments next to `ctx[123]` and `123 & dirty` corresponding to the old ctx key names, so that the code isn't so mysterious looking. Is this something it would be possible to add without too many changes? If the code that generates these is fairly localized, it seems like hopefully this would mostly be a question of whether code-red can support something like that.

Rich-Harris commented:
Yes. I've pushed a change to that effect, but... well:

```diff
-if (dirty & 2) set_data(t4, ctx[1]);
+if (dirty & /* _id9uocqrtmw00_1_ */ 2) set_data(t4, ctx[1]);
```

Not quite sure why it's not replacing the comment content — can't seem to reproduce inside code-red.

Rich-Harris commented:
okay fixed. i think this is basically done now?

Conduitry commented:
Do you think we could use comments where we're accessing `ctx[123]` as well, and not just where we're doing the bit operation stuff?

Rich-Harris commented:
yep — done

Conduitry commented:
It looks like the failing tests are from precisely those mysterious changes to style-related things in a few tests. It seems like something weird happened when you ran the tests locally to update the expected js samples? I have no idea.

Conduitry commented:
Pushed an update. I'm at a loss as to what could've happened. The tests pass locally for me now - hopefully they will here as well.

Conduitry commented:
#3977 was actually responsible for the weird `expected.js` changes here. If the very first thing the compiler compiles doesn't have component styles, weird stuff happens. During CI, this is normally hidden, since `css` is alphabetically first, and so the first test that is run does have component styles.

I bet when you were running the tests locally just before committing the changes to the `expected.js` files, you using `describe.only` to only run the `js` tests, or were otherwise doing something that caused the first compiled component to not have component styles.

Rich-Harris commented:
Good news — I just rebuilt svelte.dev with this branch, and the combined weight of all the minified JS chunks is 3.8% smaller. Zipped, the difference is less (1.3%), but still not nothing. Given how much of the weight of the site is accounted for by CodeMirror, I'm encouraged by these numbers. 

I haven't done any performance testing, but based on microbenchmarks of object lookups vs bitwise operations, this is basically guaranteed to be faster, especially since we're no longer constantly changing the shape of the `changed` object. And each component will use slightly less memory as well. Win all round.

I'll merge this in, and then maybe start testing another hypothesis: since it's now nice and easy to say 'everything changed' (by setting the bitmask to `-1`), it may be possible to unify initial creation and update, which would remove a lot of duplicated code. Not *totally* straightforward — need to think about transitions, and how to create a component without running its `<script>` block until it first gets props — but worth exploring.

RedHatter commented:
These changes look great! Smaller bundle size and better performance. They do cause an issue with svelte-devtools though. Now that the ctx object is a simple array I have no way to map the values back to their original names. For example, previously a user might see something like this

```
State
string: "testing"
number: 5
```
Now in 3.16 they would instead see
```
State
0: "testing"
1: 5
```

I'm not sure how to solve this. Could some sort of ctx index to local name map be exposed in dev mode, similar to how the props work?

Devtools is still "usable" you can still see and change the state it's just completely unnamed. Let me know if you have any ideas.

Rich-Harris commented:
@RedHatter I think that's the right approach, yeah (a lookup provided in dev mode). Mind raising an issue?

(for future reference: I don't get email notifications of activity on issues and PRs, so comments on stuff that's closed/merged are very likely to pass me by. I only happened to see this because of an ancient open tab 😀 )
